datalake/src/main/java/com/sequenceiq/datalake/flow/freeipa/upscale/FreeIpaUpscaleActions.java

```diff
@@ -0,0 +1,148 @@
+package com.sequenceiq.datalake.flow.freeipa.upscale;
+
+import static com.sequenceiq.datalake.flow.freeipa.upscale.FreeIpaUpscaleEvent.FREE_IPA_UPSCALE_FAILED_EVENT;
+import static com.sequenceiq.datalake.flow.freeipa.upscale.FreeIpaUpscaleEvent.FREE_IPA_UPSCALE_FINALIZED_EVENT;
+import static com.sequenceiq.datalake.flow.freeipa.upscale.FreeIpaUpscaleEvent.FREE_IPA_UPSCALE_IN_PROGRESS_EVENT;
+import static com.sequenceiq.datalake.flow.freeipa.upscale.FreeIpaUpscaleEvent.FREE_IPA_UPSCALE_SKIPPED_EVENT;
+
+import java.util.Map;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.statemachine.StateContext;
+import org.springframework.statemachine.action.Action;
+
+import com.sequenceiq.datalake.flow.SdxContext;
+import com.sequenceiq.datalake.flow.SdxFailedEvent;
+import com.sequenceiq.datalake.flow.freeipa.upscale.event.FreeIpaUpscaleFailedEvent;
+import com.sequenceiq.datalake.flow.freeipa.upscale.event.FreeIpaUpscaleStartEvent;
+import com.sequenceiq.datalake.flow.freeipa.upscale.event.FreeIpaUpscaleSuccessEvent;
+import com.sequenceiq.datalake.flow.freeipa.upscale.event.FreeIpaUpscaleWaitRequest;
+import com.sequenceiq.datalake.metric.SdxMetricService;
+import com.sequenceiq.datalake.service.AbstractSdxAction;
+import com.sequenceiq.datalake.service.FreeipaService;
+import com.sequenceiq.flow.core.Flow;
+import com.sequenceiq.flow.core.FlowEvent;
+import com.sequenceiq.flow.core.FlowParameters;
+import com.sequenceiq.flow.core.FlowState;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.AvailabilityType;
+
+@Configuration
+public class FreeIpaUpscaleActions {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FreeIpaUpscaleActions.class);
+
+    @Inject
+    private FreeipaService freeipaService;
+
+    @Inject
+    private SdxMetricService metricService;
+
+    @Bean(name = "FREE_IPA_UPSCALE_START_STATE")
+    public Action<?, ?> freeIpaUpscale() {
+        return new AbstractSdxAction<>(FreeIpaUpscaleStartEvent.class) {
+            @Override
+            protected SdxContext createFlowContext(FlowParameters flowParameters, StateContext<FlowState, FlowEvent> stateContext,
+                    FreeIpaUpscaleStartEvent payload) {
+                return SdxContext.from(flowParameters, payload);
+            }
+
+            @Override
+            protected void doExecute(SdxContext context, FreeIpaUpscaleStartEvent payload, Map<Object, Object> variables) {
+
+                int maxCount = AvailabilityType.HA.getInstanceCount();
+                if (maxCount > freeipaService.getNodeCount(payload.getEnvCrn())) {
```

Comment:
The check here should stay in case there is another usage of this flow, instead of handling an error when the upscale call fails. The flowchain is doing the preemptively.